pkg/upload/dsym.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/bugsnag/bugsnag-cli/pkg/log"
+	"github.com/bugsnag/bugsnag-cli/pkg/server"
+	"github.com/bugsnag/bugsnag-cli/pkg/utils"
+)
+
+type Dsym struct {
+	VersionName string      `help:"The version of the application."`
+	Scheme      string      `help:"The name of the scheme to use when building the application."`
+	Dev         bool        `help:"Indicates whether the application is a debug or release build"`
+	DsymPath    string      `help:"Path to the dSYM" type:"path"`
+	Plist       string      `help:"Path to the Info.plist file" type:"path"`
+	ProjectRoot string      `help:"path to remove from the beginning of the filenames in the mapping file" type:"path"`
+	Path        utils.Paths `arg:"" name:"path" help:"Path to directory or file to upload" type:"path" default:"."`
+}
+
+func ProcessDsym(
+	apiKey string,
+	versionName string,
+	scheme string,
+	dev bool,
+	dsymPath string,
+	plistPath string,
+	projectRoot string,
+	paths []string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+) error {
+
+	var buildSettings *ios.XcodeBuildSettings
+	var dsyms *[]*ios.DsymFile
+	var plistData *ios.PlistData
+	var uploadOptions map[string]string
+
+	for _, path := range paths {
+		uploadInfo, err := ios.ProcessPathValue(path, projectRoot)
+		if err != nil {
+			return err
+		}
+
+		// If dsymPath is not set explicitly, use uploadInfo to set it (if available)
+		// uploadInfo.DsymPath is set when <path> is recognised as a dsym path
+		if dsymPath == "" && uploadInfo.DsymPath != "" {
+			dsymPath = uploadInfo.DsymPath
+		}
+
+		// If projectRoot is not set explicitly, use uploadInfo to set it
+		if projectRoot == "" {
+			projectRoot = uploadInfo.ProjectRoot
+		}
+
+		// If scheme is set explicitly, check if it exists
+		if scheme != "" {
+			_, err = ios.IsSchemeInPath(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+
+			// Only when the dsym path is not set, try and work out the scheme
+			if dsymPath == "" {
+				// If the scheme is not set explicitly, try to find it
+				scheme, err = ios.GetDefaultScheme(path, projectRoot)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		log.Info("Using scheme: " + utils.DisplayBlankIfEmpty(scheme))
+
+		// If the dsymPath is not fed in via <path>
+		if dsymPath == "" {
+			buildSettings, err = ios.GetXcodeBuildSettings(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+
+			// Build the dsymPath from build settings
+			dsymPath = filepath.Join(buildSettings.ConfigurationBuildDir, buildSettings.DsymName, "Contents", "Resources", "DWARF")
+
+			// Check if dsymPath exists, if not, try alternative path instead
+			err = utils.Path(dsymPath).Validate()
+			if err != nil {
+				log.Info("Could not find dSYM in expected location: " + utils.DisplayBlankIfEmpty(dsymPath))
+				dsymPath = filepath.Join(buildSettings.ConfigurationBuildDir, strings.TrimSuffix(buildSettings.DsymName, ".dSYM"))
```

Comment:
Having looked into this more I've discovered some of this is not needed. The artifact in `.app` is an iOS application bundle apparently, which when used with dwarfdump gave a valid UUID which made me think it could be used as an alternative if `.app.dSYM` is not present, but I don't think that is possible and can probably just error if it is missing.